Add the ability to have multiple confgens and nanorcs per session

philiprodrigues opened the pull request:
Implemented by parametrizing the fixtures in integrationtest_nanorc.py
"dynamically" using pytest_generate_tests. Required rearranging the
tmp file code a bit, because there were some implicit assumptions that
there would only be one json dir and one output dir. With the result
that setup_dirs became obsolete.

bieryAtFnal commented:
I've tried to make confgen_arguments and nanorc_command_list both lists of lists, but I can't seem to get that to work.  

When we do this, does the run_nanorc object become a list?  

The error messages that I see give a little hint of that:

```
ERROR four_process_quick_test.py::test_nanorc_success - TypeError: expected str, bytes or os.PathLike object, not list
ERROR four_process_quick_test.py::test_log_files - TypeError: expected str, bytes or os.PathLike object, not list
ERROR four_process_quick_test.py::test_data_file - TypeError: expected str, bytes or os.PathLike object, not list
```

philiprodrigues commented:
I don't think that `run_nanorc` should become a list - it should be a single instance of the same sort of object as it was before.

Here's the relevant part extracted from an example that works for me:

```python
# The name of the python module for the config generation
confgen_name="minidaqapp.nanorc.mdapp_multiru_gen"
# The arguments to pass to the config generator, excluding the json
# output directory (the test framework handles that)
confgen_arguments_base=[ "-d", "./frames.bin", "-o", ".", "-s", "10", "-n", "2", "-b", "1000", "-a", "1000", "--host-ru", "localhost"]
confgen_arguments = [ confgen_arguments_base, confgen_arguments_base+["--disable-trace"] ]
# The commands to run in nanorc, as a list
nanorc_command_list=[ "boot init conf start 101 wait 1 resume wait 20 pause wait 1 stop wait 2 scrap terminate".split(),
                                        "boot init conf start 101 wait 1 resume wait 10 pause wait 1 stop wait 2 scrap terminate".split() ]
```

Is that similar to what you're doing? If so, can you point me to your full example, and I'll take a look

bieryAtFnal commented:
Thanks, Phil, that helped.
I suspect that my problems were in the use of the split() function and lists of lists.

I just tried again, using a model suggested by your example, and that worked.  (Although, it seems that one can't simply add [with the plus sign] the results of two split() functions...)

In any case, my next question has to do with testing the results of the NxM different tests.  Different test conditions might result in different numbers of files or different Fragment sizes, and I'm thinking about how to set the different values that are used in the test comparisons.
I'm guessing that the main handle that we have is the `run_nanorc` object.  is that right?
Would it be possible/reasonable to look inside the test configuration directory (provided by the run_nanorc.json_dir parameter, I think) to search for strings that indicate different conditions?

Something along the lines of 

- spawn a shell command to search the json directory for the string "--enable-software-tpg", and if we find that, set the expected minimum Fragment size to 80, otherwise set the expected minimum Fragment size to 37200.

Would there be better/easier ways of doing that?
Thanks

carloschavez commented:
Hi,  I have tested this branch and I confirm that the functionality for running nanorc multiple times with different confgen parameters  and different command sequences works as expected and as explained in the README.md .

I also would like to make a suggestion for the next step of improvements to the framework. It would be useful to improve the reporting of the issues when test failed the assertion. For example when running NxM tests then only way to distinguish what nanorc run correspond to a failed test is by looking at the name reported, for example:
"test_data_file[create_json_files1-run_nanorc0]"  would correspond to  1x0  (in NxM tests)  
"test_data_file[create_json_files3-run_nanorc2]"  would correspond to  3x2  (in NxM tests) 

Or if this can't be improved from pytest,  then we should have a note in the README.md explaining how to interpret this in order to understand what tests failed.

philiprodrigues commented:
> I just tried again, using a model suggested by your example, and that worked. (Although, it seems that one can't simply add [with the plus sign] the results of two split() functions...)

Yeah, "+" appends the lists to one another, but you want another "level" of lists. ie, the second example here, not the first:

```
>>> "a b c".split() + "d e f".split()
['a', 'b', 'c', 'd', 'e', 'f']
>>> [ "a b c".split() , "d e f".split() ]
[['a', 'b', 'c'], ['d', 'e', 'f']]
```

> In any case, my next question has to do with testing the results of the NxM different tests. Different test conditions might result in different numbers of files or different Fragment sizes, and I'm thinking about how to set the different values that are used in the test comparisons.
> I'm guessing that the main handle that we have is the `run_nanorc` object. is that right?
> Would it be possible/reasonable to look inside the test configuration directory (provided by the run_nanorc.json_dir parameter, I think) to search for strings that indicate different conditions?

I've added the confgen arguments and the nanorc list of commands to the `run_nanorc` object as member variables (see README). Please take a look and let me know if that provides what you need

philiprodrigues commented:
> I also would like to make a suggestion for the next step of improvements to the framework. It would be useful to improve the reporting of the issues when test failed the assertion. For example when running NxM tests then only way to distinguish what nanorc run correspond to a failed test is by looking at the name reported, for example:
> "test_data_file[create_json_files1-run_nanorc0]" would correspond to 1x0 (in NxM tests)
> "test_data_file[create_json_files3-run_nanorc2]" would correspond to 3x2 (in NxM tests)
> 
> Or if this can't be improved from pytest, then we should have a note in the README.md explaining how to interpret this in order to understand what tests failed.

That's a good idea, and turned out to be relatively straightforward to implement. I've pushed a change to allow that. Please take a look at the README and try it out

bieryAtFnal commented:
I've made use of these latest changes in my integtests on the (new) _kbiery/multi-system-integtests_ branch in the _dfmodules_ repo.

Specifically, each "test driver" file now runs through three system configurations:  basic, with SWTPG, and with DQM.

There is logic in the tests that sense which of those three scenarios is being executed and tweaks the expected comparison values appropriately.  
To do that 'auto-sensing' of the test conditions, I check for different values in the run_nanorc.confgen_arguments lists.

I haven't done much with the "test name" functionality.  If that could be made available in a run_nanorc field, I might be tempted to do something with it.  But, that can certainly wait until later.

In general, it might be good to try to finish up our changes on Friday so that we can convince our colleagues to include them in v2.8.

bieryAtFnal commented:
I should add that the updated scripts on the kbiery/multi-system-integtests branch in the dfmodules repo make use of the latest nanorc --resume-wait/--stop-wait functionality...

philiprodrigues commented:
I can't quickly see a way to get the test name into the run_nanorc object, unfortunately, so that will have to wait. But I agree with getting this into 2.8. @bieryAtFnal @carloschavez are you both happy with merging this PR?

carloschavez commented:
Hi @bieryAtFnal @philiprodrigues  ,  Yes I am happy with merging this. 
The functionally works well.

bieryAtFnal reviewed (approved):
I too will be happy to see this merged to develop...
